itwm_example/mco/space_sampling/space_samplers.py

```diff
@@ -0,0 +1,158 @@
+import abc
+import numpy as np
+
+from traits.api import ABCHasStrictTraits, Bool, ListFloat
+from force_bdss.api import PositiveInt
+
+
+def convert_samples_pp_to_samples_total(space_dimension, nof_points):
+    samples_total = (
+        np.math.factorial(space_dimension + nof_points - 2)
+        / np.math.factorial(space_dimension - 1)
+        / np.math.factorial(nof_points - 1)
+    )
+    return int(samples_total)
+
+
+class SpaceSampler(ABCHasStrictTraits):
+    """ Base class for search space sampling from various
+    distributions.
+
+    Given the dimension of the sample vectors, and the
+    sampling resolution along each of the dimensions,
+    it provides a public method to generate a number of
+    search space samples.
+    The space search satisfies the requirement that the
+    l1-norm of all samples always equals 1.0.
+    """
+
+    #: the dimension of the sample vectors
+    dimension = PositiveInt()
+
+    #: the number of (effective) divisions along each dimension
+    resolution = PositiveInt()
+
+    def __init__(self, dimension, resolution, **kwargs):
+        super().__init__(dimension=dimension, resolution=resolution, **kwargs)
+
+    @abc.abstractmethod
+    def _get_sample_point(self):
+        pass
+
+    @abc.abstractmethod
+    def generate_space_sample(self, *args, **kwargs):
+        """ Generates specified number of search space samples
+
+        Yields
+        -------
+        generator
+            random samples of vector satisfying the
+
+        """
+        pass
+
+
+class DirichletSpaceSampler(SpaceSampler):
+    """ Search space sampler class with probability
+    distribution function defined by the Dirichlet
+    distribution [1].
+
+    The distribution is "fair", providing equal statistics
+    for each component of the sample vector. The user can
+    control how centered the distribution is, by changing
+    `alpha`:
+        - Samples are closer to bounds for alpha < 1,
+        - Sampels are concentrated in the middle of the
+        search space for alpha > 1,
+        - All samples have equal probability for alpha = 1
+
+    References
+    -------
+    [1] https://en.wikipedia.org/wiki/Dirichlet_distribution
+    """
+
+    alpha = ListFloat()
+
+    distribution_function = np.random.dirichlet
+
+    def __init__(self, dimension, resolution, alpha=None, **kwargs):
+        super().__init__(dimension, resolution, **kwargs)
+
+        if alpha is None:
+            _centering_coef = 1.0
+        else:
+            _centering_coef = alpha
+
+        self.alpha = [_centering_coef for _ in range(self.dimension)]
+
+    def _get_sample_point(self):
+        return self.distribution_function(self.alpha).tolist()
+
+    def generate_space_sample(self):
+        total_nof_points = convert_samples_pp_to_samples_total(
```

Comment:
This is a style suggestion - using `_nof_` in variable names looks a little out of place, as we would expect `_n_of_` from general naming conventions. That being said, including `_of_` in variable names also feels a little verbose, so I can see why there's a desire to combine these terms.

Here, I imagine `total_n_points` would be preferred, or even `n_points`, as in this method there are no other local variables with `points` in the name.